### PR TITLE
ASoC: soc-acpi: add snd_soc_acpi_intel_codec_search function

### PR DESCRIPTION
**Pull request by brentlu**
ASoC: soc-acpi: add snd_soc_acpi_intel_codec_search function

A new machine quirk funcion snd_soc_acpi_intel_codec_search is
implemented to detect codec type and SSP port id then pass the
information to machine driver via snd_soc_acpi_mach_params structure.

This function also generate drv_name and sof_tplg_filename based on
the codecs detected to get rid of long snd_soc_acpi_mach array for
machine driver enumeration.

To make this funcion work, ODM needs to implement a new property
"intel,ssp-port" in codec's ACPI _DSD object to indicate the SSP port
id.

Signed-off-by: Brent Lu <brent.lu@intel.com>

### CODE REVIEW

**Comment by brentlu** [comment]
Coreboot modification: add a property 'intel,ssp-port' in _DSD object of codec and amplifier device in SSDT table.

        Device (RT58)
        {
            Name (_HID, "10EC5682")  // _HID: Hardware ID
            Name (_UID, Zero)  // _UID: Unique ID
            Name (_DDN, "Headset Codec")  // _DDN: DOS Device Name
            Method (_STA, 0, NotSerialized)  // _STA: Status
            {
                Return (0x0F)
            }

            Name (_CRS, ResourceTemplate ()  // _CRS: Current Resource Settings
            {
                I2cSerialBusV2 (0x001A, ControllerInitiated, 0x00061A80,
                    AddressingMode7Bit, "\\_SB.PCI0.I2C0",
                    0x00, ResourceConsumer, , Exclusive,
                    )
                GpioInt (Edge, ActiveBoth, Exclusive, PullDefault, 0x0000,
                    "\\_SB.PCI0.GPIO", 0x00, ResourceConsumer, ,
                    )
                    {   // Pin list
                        0x0057
                    }
            })
            Name (_DSD, Package (0x02)  // _DSD: Device-Specific Data
            {
                ToUUID ("daffd814-6eba-4d8c-8a91-bc9bbf4aa301") /* Device Properties for _DSD */, 
                Package (0x03)
                {
                    Package (0x02)
                    {
                        "irq-gpios", 
                        Package (0x04)
                        {
                            \_SB.PCI0.I2C0.RT58, 
                            Zero, 
                            Zero, 
                            Zero
                        }
                    }, 

                    Package (0x02)
                    {
                        "realtek,jd-src", 
                        One
                    }, 

                    Package (0x02)
                    {
                        "intel,ssp-port", 
                        Zero
                    }
                }
            })
        }

        Device (MAXM)
        {
            Name (_HID, "MX98360A")  // _HID: Hardware ID
            Name (_UID, Zero)  // _UID: Unique ID
            Name (_DDN, "Maxim Integrated 98357A Amplifier")  // _DDN: DOS Device Name
            Method (_STA, 0, NotSerialized)  // _STA: Status
            {
                Return (0x0F)
            }

            Name (_CRS, ResourceTemplate ()  // _CRS: Current Resource Settings
            {
                GpioIo (Exclusive, PullDefault, 0x0000, 0x0000, IoRestrictionOutputOnly,
                    "\\_SB.PCI0.GPIO", 0x00, ResourceConsumer, ,
                    )
                    {   // Pin list
                        0x004B
                    }
            })
            Name (_DSD, Package (0x02)  // _DSD: Device-Specific Data
            {
                ToUUID ("daffd814-6eba-4d8c-8a91-bc9bbf4aa301") /* Device Properties for _DSD */, 
                Package (0x03)
                {
                    Package (0x02)
                    {
                        "sdmode-gpio", 
                        Package (0x04)
                        {
                            \_SB.PCI0.HDAS.MAXM, 
                            Zero, 
                            Zero, 
                            Zero
                        }
                    }, 

                    Package (0x02)
                    {
                        "sdmode-delay", 
                        0x05
                    }, 

                    Package (0x02)
                    {
                        "intel,ssp-port", 
                        0x02
                    }
                }
            })
        }

2. Use device_property_read_u32() function to read the property from machine driver.
[    5.495051] sof_rt5682 adl_mx98360_rt5682: overwrite codec ssp port 0
[    5.495073] sof_rt5682 adl_mx98360_rt5682: overwrite amp ssp port 2

File: sound/soc/intel/boards/sof_rt5682.c [hunk 1]
```diff
@@ -641,6 +661,18 @@ static struct snd_soc_dai_link *sof_card_dai_links_create(struct device *dev,
...
 	links[id].num_cpus = 1;
+
+	port = sof_rt5682_check_ssp_port_overwrite(dev, links[id].codecs);
+	if (port >= 0) {
+		dev_info(dev, "overwrite codec ssp port %d\n", port);
+		ssp_codec = port;
+	}
+
+	links[id].name = devm_kasprintf(dev, GFP_KERNEL,
+					"SSP%d-Codec", ssp_codec);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_rt5682.c]: Question: for this machine you would still need a topology that hard-codes the same value as the _DSD property, yes?

In other words, CoreBoot and topology have to match now, and the machine driver would be generic?

File: sound/soc/intel/boards/sof_rt5682.c [hunk 2]
```diff
@@ -596,6 +596,30 @@ static struct snd_soc_dai_link_component dummy_component[] = {
...
+
+	dai = snd_soc_find_dai(dlc);
+	if (!dai) {
+		dev_err(dev, "fail to get dai\n");
+		return -EINVAL;
+	}
+
+	component = dai->component;
+
+	ret = device_property_read_u32(component->dev, "intel,ssp-port", &port);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_rt5682.c]: I wonder if you need to do a get/put_device(component->dev) to prevent the codec driver from being unbound.

Comment (plbossart) [inline sound/soc/intel/boards/sof_rt5682.c]: and also maybe add a sanity check on the port?

File: sound/soc/intel/boards/sof_rt5682.c [hunk 3]
```diff
@@ -596,6 +596,30 @@ static struct snd_soc_dai_link_component dummy_component[] = {
...
+	if (ret) {
+		dev_err(component->dev, "fail to read property 'intel,ssp-port'\n");
+		return -ENODATA;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_rt5682.c]: how would this work with a kernel update where CoreBoot is not updated? You would return an error then?

**Comment by brentlu** [comment]
Update V2 version.

1. Implement snd_soc_acpi_intel_codec_list function to read property from ACPI _DSD object. Then pass the SSP port ids to machine driver via snd_soc_acpi_mach_params structure.
2. Replace the original quirk function "snd_soc_acpi_codec_list" with "snd_soc_acpi_intel_codec_list" to make it work.
3. It's backward compatible. For existing coreboot images which does not implement the "intel,ssp-port" property, machine driver will use the port ids embedded in machine driver quirk.

4. The snd_soc_acpi_intel_codec_list function will try to complete the mach->sof_tplg_filename if it's not a complete topology name. The format is like sof-< platform >-< amp name >-ssp< amp port >-< codec name >-ssp< codec port >.tplg
Ex. sof-adl-max98390-ssp2-rt5682-ssp0.tplg

File: sound/soc/intel/common/soc-acpi-intel-match-common.c [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+		ret = device_property_read_u32(dev, "intel,ssp-port", &port);
+		if (ret)
+			*ssp_port = -ENODATA;
+		else if (port > 5)
+			/* sanity check fail for port id */
+			*ssp_port = -EINVAL;
+		else
+			*ssp_port = port;
+
+		put_device(dev);
```
Comment (brentlu) [inline sound/soc/intel/common/soc-acpi-intel-match-common.c]: don't know if this put_device is necessary...any comment?

File: sound/soc/intel/common/soc-acpi-intel-match-common.c [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+		adev = acpi_dev_get_first_match_dev(list[i].acpi_hid, NULL, -1);
+		if (!adev)
+			continue;
+
+		dev = get_device(acpi_get_first_physical_node(adev));
+
+		ret = device_property_read_u32(dev, "intel,ssp-port", &port);
+		if (ret)
+			*ssp_port = -ENODATA;
+		else if (port > 5)
```
Comment (brentlu) [inline sound/soc/intel/common/soc-acpi-intel-match-common.c]: any existing macro to replace this hardcode number?

File: sound/soc/intel/boards/sof_rt5682.c (hunk 1 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_rt5682.c]: yes, we should be able to remove most of dmi quirks due to different ssp port assignment.

File: sound/soc/intel/boards/sof_rt5682.c (hunk 2 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_rt5682.c]: added in v2

Comment (brentlu) [reply sound/soc/intel/boards/sof_rt5682.c]: done

File: sound/soc/intel/common/soc-acpi-intel-adl-match.c [hunk 6]
```diff
@@ -468,9 +468,9 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_adl_machines[] = {
 	{
 		.comp_ids = &adl_rt5682_rt5682s_hp,
 		.drv_name = "adl_mx98360_rt5682",
-		.machine_quirk = snd_soc_acpi_codec_list,
+		.machine_quirk = snd_soc_acpi_intel_codec_list,
 		.quirk_data = &adl_max98360a_amp,
-		.sof_tplg_filename = "sof-adl-max98360a-rt5682.tplg",
+		.sof_tplg_filename = "sof-adl",
```
Comment (plbossart) [inline sound/soc/intel/common/soc-acpi-intel-adl-match.c]: no we can't do this. If someone does a kernel update without updating firmware/topology, then this would break.

What you are suggesting can only be done for NEW platforms. Old platforms are a done deal.

Or you have to explain how this is backwards-compatible, I don't understand what happens if there isn't any information in Coreboot, where would that ssp_codec and ssp_amp information comes from?

And I also wonder why we are not listing the SSP used for BT btw, this is another problem that remains and will still require quirks.

File: sound/soc/intel/common/soc-acpi-intel-match-common.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-match-common.c]: the number of ssp ports in known in the SOF driver. for each device we have an ssp_count.

we also pass an i2s_link_mask but I am not sure if this is set for regular Chromebooks, this was defined for es8336 codec support.

File: sound/soc/intel/common/soc-acpi-intel-match-common.c (hunk 4 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-match-common.c]: We would need to ask @andy-shev if there are any protections needed or limitation if we try to read a property from a different device.

File: include/sound/soc-acpi-intel-match.h [hunk 7]
```diff
@@ -47,4 +47,7 @@ extern struct snd_soc_acpi_mach snd_soc_acpi_intel_mtl_sdw_machines[];
  */
 extern struct snd_soc_acpi_mach snd_soc_acpi_intel_hda_machines[];
 
+/* Intel version of snd_soc_acpi_codec_list function */
+struct snd_soc_acpi_mach *snd_soc_acpi_intel_codec_list(void *arg);
```
Comment (plbossart) [inline include/sound/soc-acpi-intel-match.h]: commit message: remove the Change-Id.

File: sound/soc/intel/common/soc-acpi-intel-match-common.c [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+		if (ssp_amp >= 0)
+			len += sprintf(tplg_filename + len, "-ssp%d", ssp_amp);
+		if (idx_codec >= 0)
+			len += sprintf(tplg_filename + len, "-%s", codec_list[idx_codec].tplg_name);
+		if (ssp_codec >= 0)
+			len += sprintf(tplg_filename + len, "-ssp%d", ssp_codec);
+
+		strcat(tplg_filename, ".tplg");
+
+		mach->sof_tplg_filename = kstrdup(tplg_filename, GFP_KERNEL);
```
Comment (plbossart) [inline sound/soc/intel/common/soc-acpi-intel-match-common.c]: need to test for NULL

File: include/sound/soc-acpi-intel-match.h (hunk 7 above)
Comment (plbossart) [inline include/sound/soc-acpi-intel-match.h]: your commit message also needs to add a lot more background on what you are trying to solve, and how. The backwards-compatibility story needs to be crystal-clear.

File: sound/soc/intel/boards/sof_rt5682.c [hunk 9]
```diff
@@ -914,6 +914,17 @@ static int sof_audio_probe(struct platform_device *pdev)
 
 	ssp_codec = sof_rt5682_quirk & SOF_RT5682_SSP_CODEC_MASK;
 
+	if (mach->mach_params.ssp_valid != false) {
```
Comment (andy-shev) [inline sound/soc/intel/boards/sof_rt5682.c]: Why do you need an additional variable if you have a signed integer type? I suppose that negative values there are invalid by definition, no?

File: sound/soc/intel/common/soc-acpi-intel-match-common.c [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+	struct acpi_device *adev;
+	struct device *dev;
+	int i, port, ret;
+
+	for (i = 0; i < list_len; i++) {
+		adev = acpi_dev_get_first_match_dev(list[i].acpi_hid, NULL, -1);
+		if (!adev)
+			continue;
+
+		dev = get_device(acpi_get_first_physical_node(adev));
```
Comment (andy-shev) [inline sound/soc/intel/common/soc-acpi-intel-match-common.c]: Why do you need `get_device()`? The `acpi_dev_get_first_match_dev()` bumps a reference count on the ACPI companion device. and as long as it's bumped, the physical device won't gone. This is my understanding, but I didn't look into that for a long time, something might be changed. Hence, please double check it.

File: sound/soc/intel/common/soc-acpi-intel-match-common.c [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+		ret = device_property_read_u32(dev, "intel,ssp-port", &port);
```
Comment (andy-shev) [inline sound/soc/intel/common/soc-acpi-intel-match-common.c]: Where is this property being documented? We really need something like devicetree bindings analogue for the ACPI specific device properties, so we won't reinvent the wheel here and there.

File: sound/soc/intel/common/soc-acpi-intel-match-common.c (hunk 4 above)
Comment (andy-shev) [reply sound/soc/intel/common/soc-acpi-intel-match-common.c]: @plbossart In this case it's bound device, and my understanding is described above.

File: sound/soc/intel/common/soc-acpi-intel-match-common.c [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+		if (idx_amp >= 0)
+			len += sprintf(tplg_filename + len, "-%s", amp_list[idx_amp].tplg_name);
+		if (ssp_amp >= 0)
+			len += sprintf(tplg_filename + len, "-ssp%d", ssp_amp);
+		if (idx_codec >= 0)
+			len += sprintf(tplg_filename + len, "-%s", codec_list[idx_codec].tplg_name);
+		if (ssp_codec >= 0)
+			len += sprintf(tplg_filename + len, "-ssp%d", ssp_codec);
+
+		strcat(tplg_filename, ".tplg");
```
Comment (andy-shev) [inline sound/soc/intel/common/soc-acpi-intel-match-common.c]: Oh la la,  can these `snprintf()`:s be simplified somehow?

File: sound/soc/intel/common/soc-acpi-intel-match-common.c (hunk 8 above)
Comment (andy-shev) [reply sound/soc/intel/common/soc-acpi-intel-match-common.c]: @plbossart  ...or rather think how `kasprintf()` can be used.

**Review by andy-shev — CHANGES_REQUESTED** [review]

File: sound/soc/intel/common/soc-acpi-intel-match-common.c (hunk 10 above)
Comment (brentlu) [reply sound/soc/intel/common/soc-acpi-intel-match-common.c]: acpi_dev_get_first_match_dev() calls bus_find_device() and get_device() is called in bus_find_device() so the get_device() here seems to be redundant. Thanks.

File: sound/soc/intel/common/soc-acpi-intel-adl-match.c (hunk 6 above)
Comment (brentlu) [reply sound/soc/intel/common/soc-acpi-intel-adl-match.c]: Sorry for the confusing. This is just my test code to see if it works on brya board. I think we could introduce it to MTL Chrome projects. Intel BT is enumerated from USB so I'm afraid there is no proper place to add this property for BT SSP port id.

File: sound/soc/intel/boards/sof_rt5682.c (hunk 9 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_rt5682.c]: remove this variable

**Comment by brentlu** [comment]
1. Only for boards starting from MTL, nothing change to ADL and prior platforms
2. ODM needs to add "intel,ssp-port" property to codec's ACPI _DSD object or enumeration will fail
3. Only one board (i.e. mtl_acpi_rt5682) is added to each machine driver for all boards with specific headphone codec (i.e. rt5682) with/without any codec amplifier on specific platform (i.e. mtl). No longer add lots of board configuration to machine driver for each codec/amplifier combination
4. The machine quirk function also generate TPLG file name according to the headphone/amplifier detected. It seems to me that adding port id to the naming convention (i.e. sof-adl-max98360a-ssp1-rt5682-ssp0.tplg) would be a good idea.
5. Intel BT is enumerated from USB so I find no place to add "intel,ssp-port" property. On MTL only SSP1 and SSP2 are available for BT offload and Google insists to use SSP1 for SDW device on reference design. So our code will try to use SSP1 for BT offload first as long as it's not occupied.

**Comment by plbossart** [comment]
@brentlu @cujomalainey @bardliao I don't see where this PR is going, I am tempted to close unless there's a burning desire and urgency to change the way the machine drivers are integrated.

**Comment by cujomalainey** [comment]
I can see the appeal here of having DAI come from coreboot, especially with how many quirks we put in TGL. But I think it would only be worth it if we could also forward this information to the topology and remove the config there as well

**Comment by plbossart** [comment]
> I can see the appeal here of having DAI come from coreboot, especially with how many quirks we put in TGL. But I think it would only be worth it if we could also forward this information to the topology and remove the config there as well

Fully agree, we're missing a 'hw configuration' provided by platform firmware to the topology and machine driver. We keep re-adding information and hard-code it and that's clearly getting in the way of all the hardware reworks and re-configurations. I wanted to e.g. change which SSP is used on multiple skews, I could never break the topology into submission. Just horrible.

**Comment by cujomalainey** [comment]
> Fully agree, we're missing a 'hw configuration' provided by platform firmware to the topology and machine driver. We keep re-adding information and hard-code it and that's clearly getting in the way of all the hardware reworks and re-configurations. I wanted to e.g. change which SSP is used on multiple skews, I could never break the topology into submission. Just horrible.

IIRC we want to avoid NHLT issues as well. What i think we need to be wary of is putting to much information in the table. Just enough to figure out the mapping. I would think things like clocking can still be in the topology. Otherwise we might end up in quirk hell again for the opposite side of the problem.

**Comment by plbossart** [comment]
> > Fully agree, we're missing a 'hw configuration' provided by platform firmware to the topology and machine driver. We keep re-adding information and hard-code it and that's clearly getting in the way of all the hardware reworks and re-configurations. I wanted to e.g. change which SSP is used on multiple skews, I could never break the topology into submission. Just horrible.
> 
> IIRC we want to avoid NHLT issues as well. What i think we need to be wary of is putting to much information in the table. Just enough to figure out the mapping. I would think things like clocking can still be in the topology. Otherwise we might end up in quirk hell again for the opposite side of the problem.

Yes, the mapping is the most useful information.
The clocking could indicate restrictions, e.g. some codecs only want multiples of 16 bit slots. But the details of the clocks (bclock rate, number of slots, etc) could be selected in topology based on those restrictions. It's different to adding all possible configurations as done for NHLT, that was clearly one bridge too far.

**Comment by plbossart** [comment]
@brentlu @cujomalainey  I'll close this PR as it's been 6 months since the start and we don't have a clear direction. If you still want this to happen, please resubmit a new PR. thanks!
